src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +624,231 @@ internal SafeFileHandle OpenPort(string portName)
+                Interop.Sys.PollEvents events = PollEvents(1,
+                                                           pollReadEvents: _readQueue.Count > 0,
+                                                           pollWriteEvents: _writeQueue.Count > 0);
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLNVAL) ||
+                    events.HasFlag(Interop.Sys.PollEvents.POLLERR))
+                {
+                    // bad descriptor or some other error we can't handle
+                    break;
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLIN))
+                {
+                    int bytesRead = DoIORequest(_readQueue, ProcessRead);
+                    totalBytesRead += bytesRead;
+                }
+
+                // check if there is any new data (either already read or in the driver input)
+                // this event is private and handled inside of SerialPort
+                // which then throttles it with the threshold
+                long totalBytesAvailable = totalBytesRead + BytesToRead;
+                if (totalBytesAvailable > lastTotalBytesAvailable)
+                {
+                    lastTotalBytesAvailable = totalBytesAvailable;
+
+                    // We need new task so that this thread doesn't get deadlocked when someone calls
+                    // Read from within the event
+
+                    Task.Factory.StartNew(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Chars));
+                        },
+                        this,
+                        CancellationToken.None,
+                        TaskCreationOptions.DenyChildAttach,
+                        TaskScheduler.Default);
+                }
+
+                if (events.HasFlag(Interop.Sys.PollEvents.POLLOUT))
+                {
+                    DoIORequest(_writeQueue, ProcessWrite);
+                }
+            }
+        }
+
+        private static CancellationToken GetCancellationTokenFromTimeout(int timeoutMs)
+        {
+            return timeoutMs == SerialPort.InfiniteTimeout ?
+                CancellationToken.None :
+                (new CancellationTokenSource(Math.Max(timeoutMs, TimeoutResolution))).Token;
+        }
+
+        class SerialStreamAsyncResult
+        {
+            public Memory<byte> Buffer { get; private set; }
+            public bool IsCompleted => _tcs.Task.IsCompleted;
+            public Task<int> UnderlyingTask => _tcs.Task;
+
+            private TaskCompletionSource<int> _tcs = new TaskCompletionSource<int>();
+            private CancellationToken _cancellationToken;
+
+            public SerialStreamAsyncResult(CancellationToken ct, Memory<byte> buffer)
+            {
+                _cancellationToken = ct;
+                ct.Register(() => _tcs.TrySetCanceled());
```

Comment:
This allocates a closure and a delegate.  It'd be better as:
```C#
ct.Register(s => ((TaskCompletionSource<int>)s).TrySetCanceled(), _tcs);
```

Additionally, and more importantly, this registration is getting leaked into the input CancellationToken.  Register returns a CancellationTokenRegistration which should be stored and Dispose'd of when the operation completes.

Reply:
this sounds like it could be done by compiler/JIT but will change it since they don't do it